File: src/wire.js
export function decodeSnapshot(bytes) {
  const d = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  let o = 0;
  const u8 = () => d.getUint8(o++),
    i8 = () => d.getInt8(o++),
    u16 = () => {
      const n = d.getUint16(o, true);
      o += 2;
      return n;
    },
    i16 = () => {
      const n = d.getInt16(o, true);
      o += 2;
      return n;
    },
    u32 = () => {
      const n = d.getUint32(o, true);
      o += 4;
      return n;
    },
    f = () => {
      const n = d.getFloat32(o, true);
      o += 4;
      return n;
    },
    q = () => i16() / 1000;
  if (u8() !== 83 || u8() !== 66 || u8() !== 50 || u8() !== 8)
    throw new Error("Server version mismatch. Restart the game server.");
  const s = { tick: u32(), time: f(), period: u8() },
    flags = u8(),
    team = u8();
  s.over = !!(flags & 1);
  s.restartPhase = (flags >> 2) & 3;
  s.controlled = [u8(), u8()];
  s.score = [u16(), u16()];
  s.charge = [q(), q()];
  s.pause = q();
  s.stars = [u8(), u8()];
  s.multiplier = i8();
  s.logicalView = [u16(), u16()];
  const medicalCode = u8(),
    origin = [i16(), i16()],
    medics = [
      [i16(), i16()],
      [i16(), i16()],
    ];
  if (medicalCode) {
    const player = medicalCode & 31,
      phase = (medicalCode >> 5) - 1;
    if (player >= 18 || phase < 0 || phase > 3)
      throw new Error("Invalid medical state.");
    s.medical = { player, phase, origin, medics };
  }

  s.ball = {
    x: f(),
    z: f(),
    h: f(),
    vx: f(),
    vz: f(),
    vh: f(),
    owner: i8(),
    lastTouch: i8(),
  };
  s.event = {
    id: u32(),
    kind: u8(),
    actor: i8(),
    target: i8(),
    x: f(),
    z: f(),
    h: f(),
  };
  s.players = Array.from({ length: 18 }, (_, i) => ({
    x: i16() / 100,
    z: i16() / 100,
    fx: q(),
    fz: q(),
    stun: q(),
    action: u8(),
    actionTime: q(),
    cooldown: q(),
    team: Math.floor(i / 9),
  }));
  const electric = u8();
  s.ball.electric = electric & 0x7f;
  s.ball.charged = (electric & 0x80) !== 0;
  s.effect = { kind: u8(), team: i8(), time: q() };
  s.credits = [u16(), u16()];
  s.reserves = [u8(), u8()];
  for (const p of s.players) {
    p.health = u8();
    p.injury = q();
    p.gear = u8();
    p.stats = Array.from({ length: 8 }, u8);
  }
  s.pickups = Array.from({ length: 7 }, () => ({
    kind: u8(),
    x: q(),
    z: q(),
    wait: q(),
    life: q(),
  }));
  const eventCount = u8();
  if (eventCount > 16) throw new Error("Invalid event history length.");
  s.events = Array.from({ length: eventCount }, () => ({
    id: u32(),
    kind: u8(),
    actor: i8(),
    target: i8(),
    x: f(),
    z: f(),
    h: f(),
  }));
  const str = () => {
    const n = u8(),
      v = new TextDecoder().decode(bytes.subarray(o, o + n));
    o += n;
    return v;
  };
  return {
    state: s,
    started: !!(flags & 2),
    team,
    room: str(),
    names: [str(), str()],
  };
}

File: server/wire.go
package main

import (
	"bytes"
	"encoding/binary"
	"math"
)

// Versioned, bounded binary snapshots keep all 18 players in one QUIC datagram.
func encodeSnapshot(m Snapshot) []byte {
	b := new(bytes.Buffer)
	b.Write([]byte{'S', 'B', '2', 8})
	put := func(v any) { _ = binary.Write(b, binary.LittleEndian, v) }
	u8 := func(v int) { put(uint8(v)) }
	i8 := func(v int) { put(int8(v)) }
	f := func(v float64) { put(float32(v)) }
	q := func(v float64) { put(int16(math.Round(v * 1000))) }
	str := func(s string) {
		raw := []byte(s)
		if len(raw) > 80 {
			raw = raw[:80]
		}
		u8(len(raw))
		b.Write(raw)
	}
	s := m.State
	put(uint32(s.Tick))
	f(s.Time)
	u8(s.Period)
	flags := 0
	if s.Over {
		flags |= 1
	}
	if m.Started {
		flags |= 2
	}
	u8(flags)
	u8(m.Team)
	u8(s.Controlled[0])
	u8(s.Controlled[1])
	for _, n := range s.Score {
		put(uint16(n))
	}
	q(s.Charge[0])
	q(s.Charge[1])
	q(s.Pause)
	u8(int(s.Stars[0]))
	u8(int(s.Stars[1]))
	i8(s.Multiplier)
	put(uint16(s.logicalView[0]))
	put(uint16(s.logicalView[1]))
	medical := Medical{}
	code := 0
	if s.Medical != nil {
		medical = *s.Medical
		code = ((medical.Phase + 1) << 5) | medical.Player
	}
	u8(code)
	for _, point := range [][2]int{medical.Origin, medical.Medics[0], medical.Medics[1]} {
		put(int16(point[0]))
		put(int16(point[1]))
	}

	for _, v := range []float64{s.Ball.X, s.Ball.Z, s.Ball.H, s.Ball.VX, s.Ball.VZ, s.Ball.VH} {
		f(v)
	}
	i8(s.Ball.Owner)
	i8(s.Ball.LastTouch)
	e := s.Event
	put(e.ID)
	u8(e.Kind)
	i8(e.Actor)
	i8(e.Target)
	f(e.X)
	f(e.Z)
	f(e.H)
	for _, p := range s.Players {
		put(int16(math.Round(p.X * 100)))
		put(int16(math.Round(p.Z * 100)))
		q(p.FX)
		q(p.FZ)
		q(p.Stun)
		u8(p.Action)
		q(p.ActionTime)
		q(p.Cooldown)
	}
	electric := s.Ball.Electric
	if s.Ball.Charged {
		electric |= 0x80
	}
	u8(electric)
	u8(s.Effect.Kind)
	i8(s.Effect.Team)
	q(s.Effect.Time)
	for _, v := range s.Credits {
		put(uint16(v))
	}
	for _, v := range s.Reserves {
		u8(v)
	}
	for _, p := range s.Players {
		u8(int(math.Round(p.Health)))
		q(p.Injury)
		u8(p.Gear)
		for _, stat := range p.Stats {
			u8(stat)
		}
	}
	for _, p := range s.Pickups {
		u8(p.Kind)
		q(p.X)
		q(p.Z)
		q(p.Wait)
		q(p.Life)
	}
	u8(s.EventCount)
	for _, e := range s.Events[:s.EventCount] {
		put(e.ID)
		u8(e.Kind)
		i8(e.Actor)
		i8(e.Target)
		f(e.X)
		f(e.Z)
		f(e.H)
	}
	str(m.Room)
	str(m.Names[0])
	str(m.Names[1])
	return b.Bytes()
}
